sys/xtimer/xtimer.c

```diff
@@ -163,6 +167,9 @@ int _xtimer_msg_receive_timeout64(msg_t *m, uint64_t timeout_ticks) {
 
 int _xtimer_msg_receive_timeout(msg_t *msg, uint32_t timeout_ticks)
 {
+    if (msg_try_receive(msg) == 1) {
+        return 1;
+    }
     msg_t tmsg;
     xtimer_t t;
     _setup_timer_msg(&tmsg, &t);
```

Comment:
This doesn't fix the problem. There is still a race if the timeout callback is triggered directly after `msg_receive();` in `_msg_wait` but before `xtimer_remove(t);`. I'll update the code snippet in my other PR to show the problem.

Reply:
> There is still a race if the timeout callback is triggered directly after `msg_receive();` in `_msg_wait` but before `xtimer_remove(t);`

Why?

This fix ensures that there are no queued messages (function doesn't pass the early out). Now, if any received message was *not* a timeout message, the xtimer is removed.

In case that removed xtimer has sent a message in between (after a "real" message, and with a queue present), that could now still be queued. But the early out prevents the function from misinterpreting that message as its own (which lead to the initial issue).

> I'll update the code snippet in my other PR to show the problem.

Which one? I'm getting confused. :)

Reply:
>Now, if any received message was not a timeout message, the xtimer is removed.

Yes, but directly before the remove is executed, the timeout callback may still trigger.
So *after* `msg_receive(m);` the proper masg, but before the xtimer could be removed.
Basically directly before [this line](https://github.com/RIOT-OS/RIOT/blob/037b269a92625e2510a18aea366abbf1eecd4603/sys/xtimer/xtimer.c#L151).

> Which one? I'm getting confused. :)

[#13500 comment](https://github.com/RIOT-OS/RIOT/pull/13500#issue-380973839)

Reply:
> Yes, but directly before the remove is executed, the timeout callback may still trigger.
> So when receiving a proper msg _after_ `msg_receive(m);`, but before the xtimer could be removed.

If another "proper" message is received, no harm done. It'll queue (or let the sender wait).
If xtimer triggers before it's "remove", the remove will run empty, but that is ok.

Worst case is that the message from that timer could get queued.

Reply:
It's not about another proper message but indeed the timeout message.
One could argue that it is okay to enqueue the timeout message but I think this is nowhere in the documentation and IMO not to be preferred.

Reply:
> Worst case is that the message from that timer could get queued.

This is much easier to handle in the application. Simliar to why unix has "_EINTR"...

Reply:
But wouldn't it be much more consequent to just always return a timeout message on timeout? With a timeout message waiting in the queue the timing behavior of following calls get very "interesting" ;)

Reply:
> With a timeout message waiting in the queue the timing behavior of following calls gets very "interesting" ;)

It would receive the timeout message, but not return the timeout return value.
Hopefully, all message handlers are written like:

```
switch(msg->type):
case EXPECTED:
//...
default:
  DEBUG("unexpected message type %i\n", msg->type);
```

> But wouldn't it be much more consequent to just always return a timeout message on timeout?

You mean when a "proper" message has been received bit also the timeout triggered?
This is difficult to model without a messge queue, as the "proper" message needs to go somewhere. Also, technically, the message was received before the timeout.

Reply:
> It would receive the timeout message, but not return the timeout return value.

No. First, it would return 1, indicating a message was received. In the same step the timeout message is added to the queue. On the next call it will immediately return the timeout message without waiting for a timeout.

>You mean when a "proper" message has been received bit also the timeout triggered?

I meant that *if* we use the returned message to signal a timeout, this should be always used for indicating a timeout and not either the return value or the timeout message indicates a timeout.

Reply:
> No. First, it would return 1, indicating a message was received. In the same step the timeout message is added to the queue. On the next call it will immediately return the timeout message without waiting for a timeout.

Yes, but in the second call, it would return "1", indicating that a message was received, vs. returning "-1". The received message would be the timeout message, though.

> I meant that _if_ we use the returned message to signal a timeout,

Ah, I didn't get that _if_. Why change the API? ALso, doesn't solve the problem, as then the former queued message would be returned.

Reply:
> The received message would be the timeout message, though.

Yes, but it will return the timeout message *immediately*, that s the problem.  And also it will return the timeout message for another call of msg_recv_timeout() i.e. when another fresh timeout should actually start.

> Why change the API?

I don't want to^^ but using the message as timeout indication would be just that...

> ALso, doesn't solve the problem, as then the former queued message would be returned.

So you agree it doesn't work as is, right?